### DISCUSSION TITLE
프로젝트 성능개선 경험

### DISCUSSION THREAD

**Discussion by michoball** [wanted-9th-3team/frontend-study#17] [프론트 기술 질문 공간]
프로젝트를 하시면서 성능을 개선해본 경험이 있으신가요? 왜 어떻게 했는지 알려주세요

**Comment by michoball** [comment]
kakao map api에 있는 overlay 기능을 프로젝트에 맞게 개선하여 사용한 경험이 있습니다.

kakao map api 에서 제공하는 overlay 기능은 dom element 나 HTML 문자열만 인식할 수 있어서 원하는 모습의 overlay를 만들기 위해 직접 탬플릿 리터럴로 하드코딩을 하여 넣는 방법밖에 없었습니다. 

이를 개선하여 react component로 overlay를 만들기 위해 react 의 createPortal을 사용하였습니다. 
 kakao overlay기능에는 특수한 id 값을 가지는 div만 생성하고  createPortal로 해당 id값을 가지는 div로 새로운 portal을 열어서 react component로 만든 customOverlay 컴포넌트를 연결했습니다. 

이러한 개선을 통해 overlay 에 들어갈 정보들을 props로 전달 받거나 전역 상태 값 selector로 받아올 수 있게 되었고 하드코딩된 스타일과 HTML 코드 보다 디버깅이 용이해졌습니다.

**Comment by WooGie911** [comment]
find apple 이라는 애플 상품 중고거래 플랫폼 프로젝트에서 이미지 압축 라이브러리를 사용하여 성능을 개선한 경험이 있습니다.

중고거래 특성상 상품을 등록하는 과정에서 업로드 해야하는 이미지 크기가 큰 경우 너무 오래걸리거나 상품 등록이 안되는 문제를 해결하기 위해 Browser Image Compression 라는 이미지 압축 라이브러리를 사용하였습니다. 

기존에 이미지 다중 업로드를 위해 이미지 업로드 훅을 만들어 사용중이었기에 이를 수정하여 압축을 할것인가, 어떤 크기로 할것인가, 몇장까지 업로드 할수 있게 할것인가 등 옵션을 매개변수로 받아 사용함으로서 가독성과 편의성 또한 증가하였습니다.

이러한 개선을 통해 업로드가 되지 않는 문제를 해결함과 동시에 업로드 시간도 줄이는 효과를 얻은 경험이 있습니다.

**Comment by bona373737** [comment]
react프로젝트를 개발 후 배포하던 과정에서 첫로딩시간이 3초 가량 긴시간이 소요되어 개선이 필요했던 경험이 있습니다. 
당시 React.lazy함수를 사용하여 필요한 컴포넌트만 먼저 로딩되도록 수정하여 프로젝트의 첫로딩시간을 1초 단축 시킬 수 있었습니다.
